### PR TITLE
Update config documentation

### PR DESCRIPTION
**Pull request by eerovaher**
### Description

I have updated the documentation of the configuration system. I've added hyperlinks, improved formatting and phrasing and added examples. But perhaps most importantly I have enabled doctests for many code blocks. Previously none of the code in the docs was tested, but now at least the examples of interacting with the configuration system at runtime are covered by tests.

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Do the proposed changes actually accomplish desired goals?
- [x] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [x] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [x] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [x] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [x] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label.
- [x] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [x] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [x] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

**Comment by pllim** [comment]
Thanks! It is in my todo list to review this but I have no time estimate.

File: docs/config/index.rst [hunk 1]
```diff
@@ -195,7 +215,7 @@ Exploring Configuration
 To see what configuration parameters are defined for a given ``conf``::
 
     >>> from astropy.utils.iers import conf
-    >>> [key for key in conf]
+    >>> list(conf)
```
Comment (pllim) [inline docs/config/index.rst:218]: Good catch!

File: docs/config/index.rst [hunk 2]
```diff
@@ -205,20 +225,22 @@ To see what configuration parameters are defined for a given ``conf``::
 
 You can also iterate through ``conf`` in a dictionary-like fashion::
 
-    >>> [key for key in conf.keys()]
+    >>> list(conf.keys())
```
Comment (pllim) [inline docs/config/index.rst:228]: As in Python `dict`, `.keys()` is redundant, as shown above.
```suggestion
    >>> list(conf)
```

File: docs/config/index.rst [hunk 3]
```diff
@@ -205,20 +225,22 @@ To see what configuration parameters are defined for a given ``conf``::
...
     [<ConfigItem: name='auto_download' value=True at ...>,
      <ConfigItem: name='auto_max_age' value=30.0 at ...>,
      ...,
      <ConfigItem: name='ietf_leap_second_auto_url' value=...>]
     >>> for (key, cfgitem) in conf.items():
     ...     if key == 'auto_max_age':
     ...         print(f'{cfgitem.description} Value is {cfgitem()}')
-    Maximum age (days) of predictive data before auto-downloading. Default is 30. Value is 30.0
+    Maximum age (days) of predictive data before auto-downloading. See "Auto
+    refresh behavior" in astropy.utils.iers documentation for details.Default
```
Comment (pllim) [inline docs/config/index.rst:242]: The lack of whitespace after period is a bug if it is indeed what is printed on screen.
```suggestion
    refresh behavior" in astropy.utils.iers documentation for details. Default
```

**Review by pllim — COMMENTED** [review]
Overall, very nice update. Thanks!

Since you have enabled doctest here, I think it is important to run an astropy config reset at the end of this page using `.. testcleanup::` as documented in https://www.sphinx-doc.org/en/master/usage/extensions/doctest.html

File: docs/config/index.rst (hunk 3 above)
Comment (eerovaher) [reply docs/config/index.rst:242]: There really is no whitespace: https://github.com/astropy/astropy/blob/23404eeb2caeea315cea89b6cb8b9b2fee4f13be/astropy/utils/iers/iers.py#L121-L125
Should I fix it here or would a separate pull request be more appropriate?

File: docs/config/index.rst (hunk 2 above)
Comment (eerovaher) [reply docs/config/index.rst:228]: This entire line is redundant, so I'll delete it.

File: docs/config/index.rst [hunk 4]
```diff
@@ -9,42 +7,41 @@ Configuration System (`astropy.config`)
...
-    >>> create_config_file('astropy')
+    >>> create_config_file('astropy')  # doctest: +SKIP
 
-You can find the exact location of this file by doing::
+The exact location of this file can be obtained with
+:func:`~astropy.config.get_config_dir`::
 
     >>> from astropy.config import get_config_dir
-    >>> get_config_dir()
+    >>> get_config_dir()  # doctest: +SKIP
```
[the remark this answers is not in the document]
Comment (eerovaher) [reply docs/config/index.rst:35]: The output of this function is not being tested, but at least its existence is.

File: docs/config/index.rst (hunk 3 above)
Comment (pllim) [reply docs/config/index.rst:242]: Personally, I think it is trivial enough to tack on here, but since this is IERS, let's ask @mhvk or @eteq .

Comment (mhvk) [reply docs/config/index.rst:242]: Definitely just go ahead and fix it here! 

Aside: thanks for yet another great set of clean-ups!

**Comment by eerovaher** [comment]
The documentation build fails with `WARNING: Unknown directive type "testcleanup".` and I don't understand why. Everything I've tried has resulted in the same outcome.  This directive is not used at all in `astropy`, so there's no good example I could follow. The commit message of 38dc883cad0e2bdb55e77951833b664c48ed3b95 suggests that this isn't just my problem.

At the moment the `testcleanup` is not needed because the configuration is left in a state that allows all following tests to pass, but guaranteeing that with `testcleanup` would be preferable.

**Comment by pllim** [comment]
Well, that is unfortunate! Looks like `testcleanup` needs `sphinx.ext.doctest` but we use `pytest_doctestplus.sphinx.doctestplus` (inherited from `sphinx-astropy`), which only defines `testsetup` but not `testcleanup`! @saimn or @bsipocz , do you remember why this is the case?

https://github.com/astropy/pytest-doctestplus/blob/16d6d4527439516d5a370b34941e4f1f549268f5/pytest_doctestplus/sphinx/doctestplus.py#L54

**Comment by pllim** [comment]
@eerovaher , let's see if this can be fixed upstream in `pytest-doctestplus` in a week or two. If not, we can proceed without `testcleanup` and open a follow-up issue. Thanks for your patience!

**Comment by pllim** [comment]
I have a patch upstream at astropy/pytest-doctestplus#165 . If that is accepted, should not be hard to get a new `pytest-doctestplus` release out and then use `testcleanup` here.

**Comment by eerovaher** [comment]
I'll keep an eye out for the release of the next version of `pytest-doctestplus`, and I can squash the last two commits here to restart the tests.

**Comment by pllim** [comment]
`pytest-doctestplus` 0.11.0 is released. 🤞

**Comment by pllim** [comment]
I am going to close/re-open to re-trigger RTD

**Comment by pllim** [comment]
Hmm the RTD build is taking super long... it might time out.

**Comment by eerovaher** [comment]
The documentation build seems to have succeeded, but this pull request should be rebased after a  new version of `pytest-astropy` has been released and set as a requirement for `astropy`.

**Comment by pllim** [comment]
pytest-astropy v0.9.0 is now on PyPI.

**Comment by eerovaher** [comment]
Before this pull request can be merged a new release of `sphinx-astropy` is also required. See astropy/sphinx-astropy#47

**Comment by pllim** [comment]
Ops... thanks for reminding me of that one. `sphinx-astropy` v1.6.0 is also released.

**Comment by pllim** [comment]
#12187 is also merged 😸

**Review by pllim — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**